File: common/src/main/java/net/shiroha233/roadweaver/client/map/data/ClientMapNotes.java
package net.shiroha233.roadweaver.client.map.data;

import net.minecraft.core.BlockPos;
import net.minecraft.client.Minecraft;
import net.minecraft.resources.ResourceLocation;
import net.shiroha233.roadweaver.map.search.MapSearchResult;
import net.shiroha233.roadweaver.map.search.MapStructureSource;
import org.jetbrains.annotations.Nullable;

import java.util.*;

/**
 * 客户端地图笔记存储 - 按存档隔离并持久化
 */
public final class ClientMapNotes {
    private ClientMapNotes() {}

    private static String currentWorldId = null;

    private static final Map<MapDataStorage.DimensionPos, String> aliases = new HashMap<>();
    private static final Map<MapDataStorage.DimensionPos, List<String>> notes = new HashMap<>();

    private static boolean dirty = false;

    public static void onWorldJoin() {
        String worldId = MapDataStorage.getWorldId();
        if (worldId == null) return;

        if (worldId.equals(currentWorldId)) return;

        if (currentWorldId != null && dirty) {
            saveToFile();
        }

        currentWorldId = worldId;
        aliases.clear();
        notes.clear();
        dirty = false;

        MapDataStorage.NotesData data = MapDataStorage.loadNotes();
        for (Map.Entry<String, String> e : data.aliases.entrySet()) {
            MapDataStorage.DimensionPos key = MapDataStorage.keyToDimensionPos(e.getKey());
            if (key != null) aliases.put(key, e.getValue());
        }
        for (Map.Entry<String, List<String>> e : data.notes.entrySet()) {
            MapDataStorage.DimensionPos key = MapDataStorage.keyToDimensionPos(e.getKey());
            if (key != null) notes.put(key, new ArrayList<>(e.getValue()));
        }
    }

    public static void onWorldLeave() {
        if (dirty) {
            saveToFile();
        }
        currentWorldId = null;
        aliases.clear();
        notes.clear();
        dirty = false;
    }

    public static void saveToFile() {
        MapDataStorage.NotesData data = new MapDataStorage.NotesData();
        for (Map.Entry<MapDataStorage.DimensionPos, String> e : aliases.entrySet()) {
            data.aliases.put(MapDataStorage.dimensionPosToKey(e.getKey().dimension(), e.getKey().pos()), e.getValue());
        }
        for (Map.Entry<MapDataStorage.DimensionPos, List<String>> e : notes.entrySet()) {
            data.notes.put(MapDataStorage.dimensionPosToKey(e.getKey().dimension(), e.getKey().pos()), e.getValue());
        }
        MapDataStorage.saveNotes(data);
        dirty = false;
    }

    @Nullable
    public static String getAlias(BlockPos pos) {
        MapDataStorage.DimensionPos key = currentKey(pos);
        return key == null ? null : aliases.get(key);
    }

    public static void setAlias(BlockPos pos, String alias) {
        MapDataStorage.DimensionPos key = currentKey(pos);
        if (key == null) return;
        if (alias == null || alias.isBlank()) {
            if (aliases.remove(key) != null) {
                dirty = true;
                saveToFile();
            }
        } else {
            aliases.put(key, alias);
            dirty = true;
            saveToFile();
        }
    }

    public static boolean hasAlias(BlockPos pos) {
        MapDataStorage.DimensionPos key = currentKey(pos);
        return key != null && aliases.containsKey(key);
    }

    public static List<MapSearchResult> searchAliases(String query, int limit) {
        if (query == null || query.isBlank() || limit <= 0) return List.of();
        String normalized = query.trim().toLowerCase(Locale.ROOT);
        ArrayList<MapSearchResult> results = new ArrayList<>();
        ResourceLocation dimension = currentDimension();
        for (Map.Entry<MapDataStorage.DimensionPos, String> entry : aliases.entrySet()) {
            if (dimension == null || !dimension.equals(entry.getKey().dimension())) continue;
            String alias = entry.getValue();
            if (alias == null || !alias.toLowerCase(Locale.ROOT).contains(normalized)) continue;
            results.add(new MapSearchResult(entry.getKey().pos(), alias, MapStructureSource.UNKNOWN.id()));
            if (results.size() >= limit) break;
        }
        return List.copyOf(results);
    }

    public static List<String> getNotes(BlockPos pos) {
        MapDataStorage.DimensionPos key = currentKey(pos);
        return key == null ? List.of() : notes.getOrDefault(key, List.of());
    }

    public static void addNote(BlockPos pos, String note) {
        if (note == null || note.isBlank()) return;
        MapDataStorage.DimensionPos key = currentKey(pos);
        if (key == null) return;
        notes.computeIfAbsent(key, k -> new ArrayList<>()).add(note);
        dirty = true;
        saveToFile();
    }

    public static void clearNotes(BlockPos pos) {
        MapDataStorage.DimensionPos key = currentKey(pos);
        if (key != null && notes.remove(key) != null) {
            dirty = true;
            saveToFile();
        }
    }

    public static void setNotes(BlockPos pos, List<String> noteList) {
        MapDataStorage.DimensionPos key = currentKey(pos);
        if (key == null) return;
        if (noteList == null || noteList.isEmpty()) {
            if (notes.remove(key) != null) {
                dirty = true;
                saveToFile();
            }
        } else {
            notes.put(key, new ArrayList<>(noteList));
            dirty = true;
            saveToFile();
        }
    }

    public static boolean hasNotes(BlockPos pos) {
        MapDataStorage.DimensionPos key = currentKey(pos);
        List<String> n = key == null ? null : notes.get(key);
        return n != null && !n.isEmpty();
    }

    private static MapDataStorage.DimensionPos currentKey(BlockPos pos) {
        ResourceLocation dimension = currentDimension();
        return dimension == null || pos == null ? null : new MapDataStorage.DimensionPos(dimension, pos);
    }

    private static ResourceLocation currentDimension() {
        Minecraft minecraft = Minecraft.getInstance();
        return minecraft == null || minecraft.level == null ? null : minecraft.level.dimension().location();
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/client/map/data/ClientMapNotes.kt
package net.shiroha233.roadweaver.client.map.data

import net.minecraft.core.BlockPos
import java.util.ArrayList

/**
 * 客户端地图笔记存储 - 按存档隔离并持久化
 *
 * 设计原理（参考 JourneyMap/Xaero's Map）：
 * 1. 数据存储在游戏目录下，而非存档内（玩家私人数据）
 * 2. 按存档名/服务器地址分文件夹隔离
 * 3. 进入世界时自动加载，退出/修改时自动保存
 *
 * 文件路径：.minecraft/config/roadweaver/mapdata/<存档名>/notes.json
 */
object ClientMapNotes {
    /** 当前加载的世界标识 */
    private var currentWorldId: String? = null

    /** 当前世界的数据（内存缓存） */
    private val aliases: MutableMap<BlockPos, String> = HashMap()
    private val notes: MutableMap<BlockPos, MutableList<String>> = HashMap()

    /** 是否有未保存的修改 */
    private var dirty: Boolean = false

    // ========== 世界生命周期 ==========

    /** 进入世界时调用 - 加载数据 */
    @JvmStatic
    fun onWorldJoin() {
        val worldId = MapDataStorage.getWorldId() ?: return

        // 如果是同一个世界，不重新加载
        if (worldId == currentWorldId) return

        // 保存旧世界数据
        if (currentWorldId !== null && dirty) {
            saveToFile()
        }

        // 清空并加载新世界数据
        currentWorldId = worldId
        aliases.clear()
        notes.clear()
        dirty = false

        val data = MapDataStorage.loadNotes()
        for ((k, v) in data.aliases) {
            val pos = MapDataStorage.keyToPos(k) ?: continue
            aliases[pos] = v
        }
        for ((k, v) in data.notes) {
            val pos = MapDataStorage.keyToPos(k) ?: continue
            notes[pos] = ArrayList(v)
        }
    }

    /** 退出世界时调用 - 保存数据 */
    @JvmStatic
    fun onWorldLeave() {
        if (dirty) {
            saveToFile()
        }
        currentWorldId = null
        aliases.clear()
        notes.clear()
        dirty = false
    }

    /** 保存到文件 */
    @JvmStatic
    fun saveToFile() {
        val data = MapDataStorage.NotesData()
        for ((pos, alias) in aliases) {
            data.aliases[MapDataStorage.posToKey(pos)] = alias
        }
        for ((pos, list) in notes) {
            data.notes[MapDataStorage.posToKey(pos)] = list
        }
        MapDataStorage.saveNotes(data)
        dirty = false
    }

    // ========== 别名操作 ==========

    @JvmStatic
    fun getAlias(pos: BlockPos): String? {
        return aliases[pos]
    }

    @JvmStatic
    fun setAlias(pos: BlockPos, alias: String?) {
        if (alias.isNullOrBlank()) {
            if (aliases.remove(pos) != null) {
                dirty = true
                saveToFile() // 立即保存
            }
        } else {
            aliases[pos] = alias
            dirty = true
            saveToFile() // 立即保存
        }
    }

    @JvmStatic
    fun hasAlias(pos: BlockPos): Boolean {
        return aliases.containsKey(pos)
    }

    // ========== 笔记操作 ==========

    @JvmStatic
    fun getNotes(pos: BlockPos): List<String> {
        return notes[pos] ?: emptyList()
    }

    @JvmStatic
    fun addNote(pos: BlockPos, note: String?) {
        if (note.isNullOrBlank()) return
        notes.computeIfAbsent(pos) { ArrayList() }.add(note)
        dirty = true
        saveToFile() // 立即保存
    }

    @JvmStatic
    fun clearNotes(pos: BlockPos) {
        if (notes.remove(pos) != null) {
            dirty = true
            saveToFile() // 立即保存
        }
    }

    @JvmStatic
    fun setNotes(pos: BlockPos, noteList: List<String>?) {
        if (noteList.isNullOrEmpty()) {
            if (notes.remove(pos) != null) {
                dirty = true
                saveToFile()
            }
        } else {
            notes[pos] = ArrayList(noteList)
            dirty = true
            saveToFile()
        }
    }

    @JvmStatic
    fun hasNotes(pos: BlockPos): Boolean {
        val n = notes[pos]
        return n !== null && n.isNotEmpty()
    }
}
